eng/tox/tox.ini

```diff
@@ -64,85 +67,83 @@ setenv =
   VIRTUALENV_PIP=20.3.3
   VIRTUALENV_SETUPTOOLS=67.6.0
 deps = {[base]deps}
-changedir = {toxinidir}
-install_command = python -m pip install {opts} {packages} --cache-dir {toxinidir}/../.tox_pip_cache_{envname} --ignore-installed
+install_command = python -m pip install {opts} {packages} --cache-dir {tox_root}/../.tox_pip_cache_{envname}
```

Comment:
re: https://github.com/Azure/azure-sdk-for-python/pull/30159#issuecomment-1542755675

> That's why you notice stuff like a different pip cache directory for each tox environment, etc.

If that's the goal, is the `--cache-dir` in the `install_command` not problematic:


Consider:

```
sdk/core/
├── azure/
├── azure-common/
├── azure-core/
├── azure-core-experimental/
├── azure-core-tracing-opencensus/
├── azure-core-tracing-opentelemetry/
├── azure-mgmt/
├── azure-mgmt-core/
├── azure-servicemanagement-legacy/
├── ci.yml
└── conda-recipe/
```

Running the following commands in parallel will end up using the same pip cache `sdk/core/.tox_pip_cache_whl`:

* `tox run -e whl --root sdk/core/azure-mgmt-core`
* `tox run -e whl --root sdk/core/azure-core`

As far as I'm aware, this PR hasn't changed what the `cache-dir` path was.

---


Using an environment specific path like `{envtmpdir}/.cache/pip` would resolve to different locations (`sdk/core/azure-mgmt-core/.tox/whl/.cache/pip`, `sdk/core/azure-core/.tox/whl/.cache/pip`)

Reply:
> Running the following commands in parallel will end up using the same pip cache sdk/core/.tox_pip_cache_whl:

Sure, but we only run one package at a time, all _environments_ for that package in parallel. So the `envname` would be different for each, and won't stomp on each other.

Reply:
And reasoning behind placing the pip cache in the service folder? Each subsequent invocation of a given environment name will be sped up because the downloaded reqs will already be present in the cache folder.

This basically means that after the first run `whl`, each subsequent package that invokes `whl` won't need to re-download resources.

Reply:
I responded to this comment yesterday, but now that response is gone. I don't know where it went 👎 

- We put a cache dir in the `service` directory for each tox environment.
- Within CI, we discover the targeted packages, figure out which tox envs are skipped, etc, and come up with a set of `tox` invocations. One for each package in the service that is triggering CI.
- Invoke all tox environments for a given package in parallel, iterating across all packages that are contained within a service directory.
- This means that sharing at the `service` level on a `per environment` basis, we avoid parallel access issues while _also_ ensuring that we don't download fresh dependencies on each tox environment (past the first one).